File: scripts/templates/mobile/compare.py
"""⑥ 比較 — `/m/{ja,en}/compare` per MOBILE_DESIGN.md §5.5.

Static shell + compare island. URL params `?a=<id>&b=<id>` select the
two occupations. Island fetches /data.detail/<padded>.json for each
plus /data.profile5.json once for radar data.
"""
from __future__ import annotations

from html import escape

from lib.i18n import t, t_plain  # type: ignore[import-not-found]
from lib.mobile_render import (  # type: ignore[import-not-found]
    Lang,
    render_footer,
    render_html_shell,
    render_menu,
    render_navbar,
    url_for,
)


SCREEN_ID = "compare"
PATH = "/compare"


def render(dictionary: dict, lang: Lang) -> str:
    title_text = t_plain(dictionary, "compare.title", lang)
    title = f"{title_text} — {t_plain(dictionary, 'brand.wordmark', lang)}"
    desc = title_text
    canonical = f"https://mirai-shigoto.com{url_for(lang, PATH)}"

    rows_legend = "".join(
        f'<tr><th class="m-cmp__row-label">{t(dictionary, f"compare.row.{k}", lang)}</th>'
        f'<td data-side="a" data-row="{k}">–</td>'
        f'<td data-side="b" data-row="{k}">–</td></tr>'
        for k in ("ai_score", "salary", "workforce", "ratio")
    )

    profile_axes = "".join(
        f'<tr><th class="m-cmp__row-label">{t(dictionary, f"detail.profile.{k}", lang)}</th>'
        f'<td data-side="a" data-axis="{k}">–</td>'
        f'<td data-side="b" data-axis="{k}">–</td></tr>'
        for k in ("creative", "social", "judgment", "physical", "routine")
    )

    main_inner = (
        f'<main class="m-frame__main m-cmp" data-island="compare">'
        f'<section class="m-cmp__hero">'
        f'<span class="m-mono-tag">{escape(t_plain(dictionary, "brand.eyebrow", lang))}</span>'
        f'<h1 class="m-cmp__title m-t-h1">'
        f'{escape(t_plain(dictionary, "compare.title", lang))}'
        f'</h1>'
        f'</section>'
        f'<div class="m-cmp__pickers">'
        f'<button class="m-cmp__picker" type="button" data-side="a">'
        f'<span class="m-mono-tag">A</span>'
        f'<span class="m-cmp__picker-name" data-role="picker-name-a">'
        f'{escape(t_plain(dictionary, "compare.placeholder.a", lang))}</span>'
        f'</button>'
        f'<span class="m-cmp__vs">vs.</span>'
        f'<button class="m-cmp__picker" type="button" data-side="b">'
        f'<span class="m-mono-tag">B</span>'
        f'<span class="m-cmp__picker-name" data-role="picker-name-b">'
        f'{escape(t_plain(dictionary, "compare.placeholder.b", lang))}</span>'
        f'</button>'
        f'</div>'
        f'<table class="m-cmp__table">'
        f'<colgroup><col class="m-cmp__col-label"><col><col></colgroup>'
        f'<tbody>'
        + rows_legend
        + f'<tr><th colspan="3" class="m-cmp__section-title">'
        f'{t(dictionary, "compare.row.profile", lang)}</th></tr>'
        + profile_axes
        + f'</tbody></table>'
        f'<p class="m-cmp__hint" data-role="hint">'
        f'?a=&lt;id&gt;&amp;b=&lt;id&gt; をURLに追加してください。例: ?a=33&amp;b=183'
        f'</p>'
        f'</main>'
    )

    body = (
        f'<div class="m-frame">'
        + render_navbar(dictionary, lang, current_path=PATH)
        + render_menu(dictionary, lang, current_screen=SCREEN_ID)
        + main_inner
        + render_footer(dictionary, lang)
        + f'</div>'
    )

    extra_head = (
        '<link rel="stylesheet" href="/styles/mobile-screen-compare.css">'
        f'<script>window.M_LANG = "{lang}";</script>'
        '<script defer src="/m/islands/compare.js"></script>'
    )

    return render_html_shell(
        lang=lang, title=title, description=desc,
        canonical=canonical, body_inner=body, extra_head=extra_head,
    )

File: m/islands/compare.js
// compare.js — vanilla JS island for ⑥ 比較.
// Reads ?a=<id>&b=<id> URL params; fetches /data.detail/<padded>.json + /data.profile5.json.
(function () {
  var lang = window.M_LANG || 'ja';
  var root = document.querySelector('[data-island="compare"]');
  if (!root) return;
  var hint = document.querySelector('[data-role="hint"]');

  function pad(n) { return String(n).padStart(4, '0'); }
  function bandClass(score) {
    if (score == null) return '';
    if (score <= 3.9) return 'm-risk-low';
    if (score <= 6.9) return 'm-risk-mid';
    return 'm-risk-high';
  }
  function fmtSalary(v) {
    if (v == null) return '–';
    return lang === 'ja' ? '¥' + v.toFixed(0) + '万' : '¥' + (v / 100).toFixed(2) + 'M';
  }
  function fmtWorkers(v) {
    if (v == null) return '–';
    return lang === 'ja' ? (v / 10000).toFixed(0) + '万人' : (v / 1000).toFixed(0) + 'k';
  }

  function fillSide(side, detail, profile) {
    var nameEl = document.querySelector('[data-role="picker-name-' + side + '"]');
    if (nameEl) {
      nameEl.textContent = lang === 'en' && detail.title.en ? detail.title.en : detail.title.ja;
    }
    var stats = detail.stats || {};
    var risk = detail.ai_risk ? detail.ai_risk.score : null;
    var rows = {
      ai_score:  risk != null ? '<span class="' + bandClass(risk) + '" style="padding:2px 8px;border-radius:999px;">' + risk + '</span>' : '–',
      salary:    fmtSalary(stats.salary_man_yen),
      workforce: fmtWorkers(stats.workers),
      ratio:     stats.recruit_ratio != null ? stats.recruit_ratio.toFixed(2) : '–',
    };
    Object.keys(rows).forEach(function (k) {
      var td = root.querySelector('td[data-side="' + side + '"][data-row="' + k + '"]');
      if (td) td.innerHTML = rows[k];
    });
    if (profile) {
      ['creative', 'social', 'judgment', 'physical', 'routine'].forEach(function (axis) {
        var v = profile[axis];
        var td = root.querySelector('td[data-side="' + side + '"][data-axis="' + axis + '"]');
        if (td) td.textContent = v != null ? v.toFixed(0) : '–';
      });
    }
  }

  var params = new URLSearchParams(location.search);
  var a = params.get('a');
  var b = params.get('b');
  if (!a || !b) return;  // hint visible by default

  if (hint) hint.style.display = 'none';

  Promise.all([
    fetch('/data.detail/' + pad(a) + '.json').then(function (r) { return r.json(); }),
    fetch('/data.detail/' + pad(b) + '.json').then(function (r) { return r.json(); }),
    fetch('/data.profile5.json').then(function (r) { return r.json(); }),
  ]).then(function (vals) {
    var detailA = vals[0], detailB = vals[1], p5 = vals[2].profiles;
    fillSide('a', detailA, p5[String(a)]);
    fillSide('b', detailB, p5[String(b)]);
  }).catch(function (e) {
    console.error('compare island error', e);
    if (hint) {
      hint.style.display = 'block';
      hint.textContent = 'データ取得エラー';
    }
  });
})();
